README.md
# receiver-controller
Radio receiver protocol translation layer for microcontrollers


immersionrc-ezuhf-433/Release_Notes.txt
# release notes

#----------------------------------------------------------------------------------------#	
██╗███╗   ███╗███╗   ███╗███████╗██████╗ ███████╗██╗ ██████╗ ███╗   ██╗██████╗  ██████╗
██║████╗ ████║████╗ ████║██╔════╝██╔══██╗██╔════╝██║██╔═══██╗████╗  ██║██╔══██╗██╔════╝
██║██╔████╔██║██╔████╔██║█████╗  ██████╔╝███████╗██║██║   ██║██╔██╗ ██║██████╔╝██║     
██║██║╚██╔╝██║██║╚██╔╝██║██╔══╝  ██╔══██╗╚════██║██║██║   ██║██║╚██╗██║██╔══██╗██║     
██║██║ ╚═╝ ██║██║ ╚═╝ ██║███████╗██║  ██║███████║██║╚██████╔╝██║ ╚████║██║  ██║╚██████╗
╚═╝╚═╝     ╚═╝╚═╝  ████████╗═██████╗  ██████╗═██╗╚═╝ ╚███████╗═╝  ╚═══╝╚═╝  ╚═╝ ╚═════╝
                   ╚══██╔══╝██╔═══██╗██╔═══██╗██║     ██╔════╝                         
                      ██║   ██║   ██║██║   ██║██║     ███████╗                         
                      ██║   ██║   ██║██║   ██║██║     ╚════██║                         
                      ██║   ╚██████╔╝╚██████╔╝███████╗███████║                         
                      ╚═╝    ╚═════╝  ╚═════╝ ╚══════╝╚══════╝                         
#----------------------------------------------------------------------------------------#	

### v1.42.5 - 04/21/2016
## What's new
- Sign CDC and Bootloader drivers with new certificate
## Bugfix
- Fix EZUHF German frequency setting.

### v1.42.4 - 12/21/2015
## Bugfix
- Fix for SUMD\SUMH\XBUS Vortex Receiver mode ( indexes were messed up ).


### v1.42.3 - 11/13/2015
## Bugfix
- Fix XuGong v2/Pro channels assigments whith none
- Fix Vortex Receiver mode compatibility
                   
				   
### v1.42.2 - 10/05/2015
## What's new
- New Control added to XuGong v2 Configure tab to configure Menu Off channel

	
### v1.42.1 - 08/31/2015
## What's new
- Speedup communication with Vortex ( read\write settings )

									
### v1.42.0 - 07/26/2015
## What's new
- Clear instruction for Vortex firmware upgrade.


### v1.41.9 - 07/06/2015
## Bugfix	
- Fix vortex communication errors.
	
	
### v1.41.8 - 07/02/2015
## What's new
- Receiver Mode control added in Vortex Configure tab ( Vortex 1.0.0.11 Beta )

### v1.41.7 - 06/09/2015
## What's new
- Extend support for Vortex
	
	
### v1.41.4 - 05/16/2015
## What's new
- Add Vortex to product list
- Configure for Vortex button added to ezUHF configuration tab 
- Add new icon for ImmersionRC
## Bugfix
- Fix wrong MessageBox title on binding success

	
### v1.41.3	- 02/02/2015
## What's new
- Add EzAntennaTracker v1/v2 
- Reorganize IRC Product list
- New icon for XuGong v2/Pro
## Bugfix
- Fix ImmersionRC drivers installation during software install
- Fix 'Configure for XuGong v2/Pro' for EzUHF-Rx

sketches/libraries/SpektrumSerialReceiver/SpektrumSerialReceiver.h
/*****************************************************************************************************************************************
 * Arduino Library for interfacing to Spektrum™ Serial Receivers.
 * Feature overview:
 *      * Reading 1024/2048 Spektrum™ serial protocols
 *      * Up to 12 RC channels + 8 X-plus channels available
 *      * Getting RX losses and bind type info
 *      * Binding Spektrum™ Serial Receivers in internal DSM2/DSMX 22ms/11ms modes
 *      * Automatic packet synchronisation
 * 
 * This source code is based on documentation obtained from:
 * https://github.com/SpektrumRC/SpektrumDocumentation/blob/master/Telemetry/Remote%20Receiver%20Interfacing.pdf
 * "Specification for Spektrum Remote Receiver Interfacing. Enabling Use of Spektrum Remotes in Third-Party Products", Rev G 2019 January 22
 * 
 * Created 20 Jun 2020
 * by Ilya Kobets (aka MrTransistor)
 * 
 * Modified 22 Jun 2020
 * by Ilya Kobets
 * 
 * Version 2.1
 *****************************************************************************************************************************************/

/**** Bind mode constants ****/
typedef enum Bind_mode{
    DSM2_22ms = 0x01,
    DSMX_22ms = 0xA2,
    DSM2_11ms = 0x12,
    DSMX_11ms = 0xB2
};
/*****************************/

/**** Main class ****/
class SpektrumSerialReceiver{
    public:
        SpektrumSerialReceiver(int bind_pin = 0);           // Constructor
        void begin(HardwareSerial& serial);                 // Initialization on HardwareSerial port. bind_pin variable should be set to receiver data pin for binding or could be left blank if not using bind functionality
        void begin(Stream& stream);                         // Initialization on Stream object
        void update();                                      // Getting and parsing data frames
        uint16_t getChannel(uint8_t chnum);                 // Returns value for specified channel
        uint8_t getRXLosses();                              // Returns number of signal losses calculated by Receiver
        Bind_mode getBindMode();                            // Returns bind mode ID, can be compared with bind mode constants to get current mode
        void bindRX(Bind_mode mode);                        // Enabling specified bind mode on receiver (uses Bind_mode constants
                                                            // bindRX() SHOULD BE USED BEFORE begin() METHOD WITHIN 200ms AFTER RECEIVER POWER-UP!!!!
    private:
        Stream* _stream;

        int _pin;

        static const uint8_t MASK_1024_CHANID  = 0xFC;      // Channel ID and servo position masks for 10 & 11 bit protocols
        static const uint8_t MASK_2048_CHANID  = 0x78;
        static const uint8_t MASK_XPLUS_CHANID = 0x03;
        static const uint8_t MASK_1024_SXPOS   = 0x03;
        static const uint8_t MASK_2048_SXPOS   = 0x07;
        static const uint8_t MASK_XPLUS_SXPOS  = 0x01;

        uint8_t packetData[14];
        uint16_t channelData[20];
        uint8_t rxlosses;
        uint8_t bindMode;

        void send_bind_pulses(uint8_t num_pulses);
};
/********************/

sketches/libraries/SpektrumSerialReceiver/SpektrumSerialReceiver.cpp
/*****************************************************************************************************************************************
 * Arduino Library for interfacing to Spektrum™ Serial Receivers.
 * Feature overview:
 *      * Reading 1024/2048 Spektrum™ serial protocols
 *      * Up to 12 RC channels + 8 X-plus channels available
 *      * Getting RX losses and bind type info
 *      * Binding Spektrum™ Serial Receivers in internal DSM2/DSMX 22ms/11ms modes
 *      * Automatic packet synchronisation
 * 
 * This source code is based on documentation obtained from:
 * https://github.com/SpektrumRC/SpektrumDocumentation/blob/master/Telemetry/Remote%20Receiver%20Interfacing.pdf
 * "Specification for Spektrum Remote Receiver Interfacing. Enabling Use of Spektrum Remotes in Third-Party Products", Rev G 2019 January 22
 * 
 * Created 20 Jun 2020
 * by Ilya Kobets (aka MrTransistor)
 * 
 * Modified 22 Jun 2020
 * by Ilya Kobets
 * 
 * Version 2.1
 *****************************************************************************************************************************************/
#include <Arduino.h>
#include "SpektrumSerialReceiver.h"

/**** Constructor ****/
SpektrumSerialReceiver::SpektrumSerialReceiver(int bind_pin=0){
    _pin = bind_pin;
}
/*********************/

/**** Initialisation on HardwareSerial ****/
void SpektrumSerialReceiver::begin(HardwareSerial& serial){
    serial.begin(115200);
    begin((Stream&)serial);
}
/******************************************/

/**** Initialisation on Stream object ****/
void SpektrumSerialReceiver::begin(Stream& stream){
    _stream = &stream;
}
/*****************************************/

/**** Updating channels ****/
void SpektrumSerialReceiver::update(){
    while (_stream->available()) {
        delay(1);
        rxlosses = _stream->read();
        bindMode = _stream->read();
        delay(1);
        for(int i = 0; i < 14; i++){
            if(!_stream->available()) return;
            packetData[i] = _stream->read();
        }
    }
    for (int i = 0; i < 7; i++) {
        if(bindMode == DSM2_22ms){
            channelData[((packetData[i * 2] & MASK_1024_CHANID) >> 2)] = ((packetData[i * 2] & MASK_1024_SXPOS) << 8) + packetData[i * 2 + 1];
        }
        else{
            if(((packetData[i * 2] & MASK_2048_CHANID) >> 3) < 12)
                channelData[((packetData[i * 2] & MASK_2048_CHANID) >> 3)] = ((packetData[i * 2] & MASK_2048_SXPOS) << 8) + packetData[i * 2 + 1];
            else{
                channelData[(12 + ((packetData[i*2] >> 1) & MASK_XPLUS_CHANID) + 4*(packetData[i*2] >> 7))] = ((packetData[i * 2] & MASK_XPLUS_SXPOS) << 8) + packetData[i * 2 + 1];
            }
        }
    }
}
/***************************/

/**** Returns specified channel value ****/
uint16_t SpektrumSerialReceiver::getChannel(uint8_t chnum){
    return channelData[chnum];
}
/*****************************************/

/**** Returns num of signal losses ****/
uint8_t SpektrumSerialReceiver::getRXLosses(){
    return rxlosses;
}
/*************************************/

/**** Returns bind mode ****/
Bind_mode SpektrumSerialReceiver::getBindMode(){
    return (Bind_mode)bindMode;
}
/***************************/

/**** Binding procedure ****/
void SpektrumSerialReceiver::bindRX(Bind_mode mode){
    pinMode(_pin, OUTPUT);
    switch(mode){
        case DSM2_22ms:
            send_bind_pulses(3);
            break;
        case DSM2_11ms:
            send_bind_pulses(5);
            break;
        case DSMX_22ms:
            send_bind_pulses(7);
            break;
        case DSMX_11ms:
            send_bind_pulses(9);
            break;
    }
    pinMode(_pin, OUTPUT);

}
/***************************/

/**** Private function for sending bind pulses ****/
void SpektrumSerialReceiver::send_bind_pulses(uint8_t num_pulses){
    for (int i = 0; i < num_pulses; i++){
        digitalWrite(_pin, 0);
        delayMicroseconds(120);
        digitalWrite(_pin, 1);
        delayMicroseconds(120);
    }
}
/**************************************************/